types/prosemirror-commands/index.d.ts

```diff
@@ -4,13 +4,26 @@
 //                 David Hahn <https://github.com/davidka>
 //                 Tim Baumann <https://github.com/timjb>
 //                 Patrick Simmelbauer <https://github.com/patsimm>
+//                 Mike Morearty <https://github.com/mmorearty>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+// TypeScript Version: 3.0
 
 import { MarkType, Node as ProsemirrorNode, NodeType, Schema } from 'prosemirror-model';
 import { EditorState, Transaction } from 'prosemirror-state';
 import { EditorView } from 'prosemirror-view';
 
+export interface Command<S extends Schema = any> {
+  (
+    state: EditorState<S>,
+    dispatch: (tr: Transaction<S>) => void,
+    view: EditorView<S>
```

Comment:
Hey! I was wondering, shouldn't it be `dispatch?` and `view?`?

Most of the command type definitions in this file use:

```TypeScript
state: EditorState<S>,
dispatch?: (tr: Transaction<S>) => void,
view?: EditorView<S>
```

and the docs say:

> A command function takes an editor state, _optionally_ a `dispatch` function that it can use to dispatch a transaction and _optionally_ an `EditorView` instance.

If we add the following code to the `prosemirror-commands-tests.ts` file, this wouldn't compile.

```TypeScript
Object.keys(commands.baseKeymap).forEach(key => {
    if (keymap[key]) {
        keymap[key] = commands.chainCommands(keymap[key], commands.baseKeymap[key]);
    } else {
        keymap[key] = commands.baseKeymap[key];
    }
});
```

(I forgot where this code came from, but apparently it's [a thing](https://github.com/search?q=Object.keys%28baseKeymap%29.forEach&type=Code))

Using the definition from this PR I get:

```
ERROR: 26:46  expect                         TypeScript@next compile error:
Argument of type 'Command<any>' is not assignable to parameter of type '(p1: EditorState<any>, p2?: ((tr: Transaction<any>) => void) | undefined, p3?: EditorView<any> | undefined) => boolean'.
  Types of parameters 'dispatch' and 'p2' are incompatible.
    Type '((tr: Transaction<any>) => void) | undefined' is not assignable to type '(tr: Transaction<any>) => void'.
      Type 'undefined' is not assignable to type '(tr: Transaction<any>) => void'.
```

If I change it to `dispatch?` and `view?` it works.

Am I doing something wrong here? Maybe I'm confusing what is a `Command` and what isn't.

Reply:
@aried3r You are right. It's important that commands can handle the case where `dispatch` and `view` are undefined, because that's used to check whether a command can be applied:

> When no dispatch callback is passed, the command should do a 'dry run', determining whether it is applicable, but not actually doing anything.

We should change that.

Reply:
@aried3r I've made a PR containing a fix: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45972